From: Casey Bodley <cbodley@redhat.com>
Date: Tue, 11 Oct 2016 15:23:16 -0400
Subject: rgw: clean up RGWShardedOmapCRManager on early return

ShardedOmapCRManager is spawning coroutines that sleep. if we don't
finish them before trying to drain_all(), they'll lead to deadlock

Fixes: http://tracker.ceph.com/issues/17571

Signed-off-by: Casey Bodley <cbodley@redhat.com>
(cherry picked from commit 2c66e9b7fad4eec017cecc6bd6005522cc0fb264)

Resolves: rhbz#1393665
---
 src/rgw/rgw_sync.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/rgw/rgw_sync.cc b/src/rgw/rgw_sync.cc
index 3d494ef..b69d329 100644
--- a/src/rgw/rgw_sync.cc
+++ b/src/rgw/rgw_sync.cc
@@ -812,6 +812,7 @@ public:
       }
       if (get_ret_status() < 0) {
         ldout(cct, 0) << "ERROR: failed to fetch metadata sections" << dendl;
+        yield entries_index->finish();
         yield lease_cr->go_down();
         drain_all();
 	return set_cr_error(get_ret_status());
@@ -827,6 +828,7 @@ public:
 	}
         if (get_ret_status() < 0) {
           ldout(cct, 0) << "ERROR: failed to fetch metadata section: " << *sections_iter << dendl;
+          yield entries_index->finish();
           yield lease_cr->go_down();
           drain_all();
           return set_cr_error(get_ret_status());
